fontforgeexe/scriptingdlg.c

```diff
@@ -189,6 +189,12 @@ return( true );
     
     if ( event->type==et_close ) {
 	SD_DoCancel( sd );
+    } else if ( event->type==et_charup ) {
+    char* contents = GGadgetGetTitle8(GWidgetGetControl(sd->gw,CID_Script));
```

Comment:
I can't take a closer look at the moment, but at a cursory glance this seems mighty inefficient. On every key press, you're reading the *full* contents of the script and performing at least one copy. You're then running over that again to find it's string length, *just* to see if it's unsaved or not.

Why can't you just always set `script_unsaved` to true? Also are you sure this is the right event to listen off?

Reply:
I skimmed `gtextfield.c` and you should be able to listen to `et_controlevent` with a subtype of `et_textchanged`. Then all you have to do is set `script_unsaved` to true in the case you receive that.

I reckon it makes more sense for that `et_save` event to also just be a subtype of the `et_controlevent` message.

Reply:
I have to know if there's text in the field, otherwise if you write a script then delete it FontForge still thinks there's one there.

Reply:
If you really want to avoid that, I can see a couple of options, extend the `et_textchanged` -> `tf_changed` event to indicate e.g. if the field is empty, or add a `GTextFieldIsEmtpy()` which does the same thing, e.g.

```c
bool GTextFieldIsEmpty(GTextField *gt) {
    return gt->text == NULL || *gt->text == '\0';
}
```

or

```c
		struct {
		    int from_pulldown;	/* -1 normally, else index into pulldown list */
		    int field_empty : 1;
		} tf_changed;

...

// In GTextFieldChanged
 e.u.control.u.tf_changed.field_empty = gt->text == NULL || *gt->text == '\0';
```